vllm/entrypoints/nixl_side_channel_server.py

```diff
@@ -0,0 +1,123 @@
+# SPDX-License-Identifier: Apache-2.0
+# SPDX-FileCopyrightText: Copyright contributors to the vLLM project
+
+import asyncio
+from typing import Any, Optional
+
+import uvicorn
+from fastapi import FastAPI, HTTPException
+
+from vllm import envs
+from vllm.config import VllmConfig
+from vllm.logger import init_logger
+
+logger = init_logger(__name__)
+
+
+class NixlSideChannelServer:
+
+    def __init__(self, vllm_config: VllmConfig, host: str, port: int):
+        self.vllm_config = vllm_config
+        self.host = host
+        self.port = port
+        self.app = FastAPI(title="vLLM NIXL Side Channel Server")
+        self.server: Optional[uvicorn.Server] = None
+        self._setup_routes()
+
+    def _setup_routes(self):
+
+        @self.app.get("/get_kv_connector_metadata")
+        @self.app.get("/get_kv_connector_metadata/{dp_rank}")
+        @self.app.get("/get_kv_connector_metadata/{dp_rank}/{tp_rank}")
+        async def get_kv_connector_metadata(dp_rank: Optional[int] = None,
+                                            tp_rank: Optional[int] = None):
+            kv_meta = self.vllm_config.parallel_config.xfer_handshake_metadata
+
+            if kv_meta is None:
+                raise HTTPException(
+                    status_code=404,
+                    detail="KV connector handshake metadata is not available")
+            if dp_rank is not None and not (dp_data := kv_meta.get(dp_rank)):
+                raise HTTPException(
+                    status_code=404,
+                    detail=f"Data parallel rank {dp_rank} not found")
+            if tp_rank is not None and dp_data is not None and not (
+                    tp_data := dp_data.get(tp_rank)):
+                raise HTTPException(
+                    status_code=404,
+                    detail=f"Tensor parallel rank {tp_rank} not found for data \
+                        parallel rank {dp_rank}")
+
+            if dp_rank is None:
+                return kv_meta
+
+            if tp_rank is None:
+                return {dp_rank: dp_data}
```

Comment:
Could avoid duplicate checks here
```suggestion
            if kv_meta is None:
                raise HTTPException(
                    status_code=404,
                    detail="KV connector handshake metadata is not available")
            if dp_rank is None:
                return kv_meta
            
            if not (dp_data := kv_meta.get(dp_rank)):
                raise HTTPException(
                    status_code=404,
                    detail=f"Data parallel rank {dp_rank} not found")

            if tp_rank is None:
                return {dp_rank: dp_data}
                
            if not (tp_data := dp_data.get(tp_rank)):
                raise HTTPException(
                    status_code=404,
                    detail=f"Tensor parallel rank {tp_rank} not found for data \
                        parallel rank {dp_rank}")
```